### PR TITLE
refactor: Improve Omnichannel queries

### PR DESCRIPTION
**Pull request by KevLehman**
## Proposed changes (including videos or screenshots)

- Move queries to models
- Correct `fields` for `projection`
- Use correct `projection` for places where we aren't using the full obj

### CODE REVIEW

File: apps/meteor/app/statistics/server/lib/statistics.ts [hunk 1]
```diff
@@ -146,14 +146,14 @@ export const statistics = {
 
 		// Number of departments
 		statsPms.push(
-			LivechatDepartment.col.count().then((count) => {
+			LivechatDepartment.col.estimatedDocumentCount().then((count) => {
```
Comment (sampaiodiego) [inline apps/meteor/app/statistics/server/lib/statistics.ts:149]: `estimatedDocumentCount` could be a `BaseRaw` method, wdyt?

File: apps/meteor/ee/app/license/server/getStatistics.ts [hunk 2]
```diff
@@ -101,28 +101,24 @@ async function getEEStatistics(): Promise<EEOnlyStats | undefined> {
 
 	statsPms.push(
 		// Total livechat monitors
-		Users.col.countDocuments({ type: 'livechat-monitor' }).then((count) => {
+		Users.col.countDocuments({ role: 'livechat-monitor' }).then((count) => {
```
Comment (sampaiodiego) [inline apps/meteor/ee/app/license/server/getStatistics.ts:104]: move this one to the model as well?

File: apps/meteor/ee/server/models/raw/LivechatDepartment.ts [hunk 3]
```diff
@@ -29,6 +30,10 @@ export class LivechatDepartmentEE extends LivechatDepartmentRaw implements ILive
 		super(db, trash);
 	}
 
+	findByParentId(parentId: string, options?: FindOptions<ILivechatDepartment>): FindCursor<ILivechatDepartment> {
+		return this.col.find({ parentId }, options);
```
Comment (sampaiodiego) [inline apps/meteor/ee/server/models/raw/LivechatDepartment.ts:34]:
```suggestion
		return this.find({ parentId }, options);
```

**Review by sampaiodiego — DISMISSED** [review]
this is great! 👏 thx a lot! 🤗 

just one overall comment, I think when using `projection` it is important to provide the correct returning type using generics.. otherwise TS will always return the full document, which can cause even more issues.

I know the generic is already there for `findOneById<T>(...)`, so for places using it it would be just a matter of providing the returning type.. but for the new methods you'd need to add the generics too.

File: apps/meteor/app/statistics/server/lib/statistics.ts (hunk 1 above)
Comment (KevLehman) [reply apps/meteor/app/statistics/server/lib/statistics.ts:149]: 🤔 that works, let me add it

File: apps/meteor/ee/server/models/raw/LivechatDepartment.ts (hunk 3 above)
Comment (KevLehman) [reply apps/meteor/ee/server/models/raw/LivechatDepartment.ts:34]: This is intentional as we're adding extra filters to the `find` method of this model. So the `col.find` is the unfiltered one

File: apps/meteor/app/livechat/server/lib/Departments.ts [hunk 4]
```diff
@@ -24,7 +24,7 @@ class DepartmentHelperClass {
 		}
 		this.logger.debug(`Department record removed: ${_id}`);
 
-		const agentsIds: string[] = await LivechatDepartmentAgents.findAgentsByDepartmentId(department._id)
+		const agentsIds: string[] = await LivechatDepartmentAgents.findAgentsByDepartmentId(department._id, { projection: { agentId: 1 } })
 			.cursor.map((agent) => agent.agentId)
 			.toArray();
```
Comment (murtaza98) [inline apps/meteor/app/livechat/server/lib/Departments.ts:29]:
```suggestion
        const agentsIds: string[] = await LivechatDepartmentAgents.findAgentsByDepartmentId<Pick<ILivechatDepartmentAgents, 'agentId'>>(
			department._id,
			{ projection: { agentId: 1 } },
		)
			.cursor.map((agent) => agent.agentId)
			.toArray();
```

File: apps/meteor/server/models/raw/LivechatDepartment.ts [hunk 5]
```diff
@@ -470,6 +477,14 @@ export class LivechatDepartmentRaw extends BaseRaw<ILivechatDepartment> implemen
 
 		return this.col.aggregate(aggregation).hasNext();
 	}
+
+	countArchived(): Promise<number> {
+		return this.col.countDocuments({ archived: true });
+	}
+
+	findByParentId(_parentId: string, _options?: FindOptions<ILivechatDepartment> | undefined): FindCursor<ILivechatDepartment> {
+		throw new Error('Method not implemented.');
```
Comment (murtaza98) [inline apps/meteor/server/models/raw/LivechatDepartment.ts:486]:
```suggestion
		throw new Error('Method not implemented in CE');
```

**Review by murtaza98 — APPROVED** [review]

File: apps/meteor/app/e2e/server/methods/setRoomKeyID.ts [hunk 6]
```diff
@@ -31,7 +31,7 @@ Meteor.methods<ServerMethods>({
 			throw new Meteor.Error('error-invalid-room', 'Invalid room', { method: 'e2e.setRoomKeyID' });
 		}
 
-		const room = await Rooms.findOneById(rid, { fields: { e2eKeyId: 1 } });
+		const room = await Rooms.findOneById(rid, { projection: { e2eKeyId: 1 } });
```
Comment (murtaza98) [inline apps/meteor/app/e2e/server/methods/setRoomKeyID.ts:34]:
```suggestion
		const room = await Rooms.findOneById<Pick<IOmnichannelRoom, '_id' | 'e2eKeyId'>>(rid, { projection: { e2eKeyId: 1 } });
```

File: apps/meteor/app/livechat/server/lib/Helper.ts [hunk 7]
```diff
@@ -549,7 +551,9 @@ export const forwardRoomToDepartment = async (room: IOmnichannelRoom, guest: ILi
 
 	const { servedBy, chatQueued } = roomTaken;
 	if (!chatQueued && oldServedBy && servedBy && oldServedBy._id === servedBy._id) {
-		const department = departmentId ? await LivechatDepartment.findOneById(departmentId) : null;
+		const department = departmentId
+			? await LivechatDepartment.findOneById(departmentId, { projection: { fallbackForwardDepartment: 1 } })
+			: null;
```
Comment (murtaza98) [inline apps/meteor/app/livechat/server/lib/Helper.ts:556]:
```suggestion
		const department = departmentId
			? await LivechatDepartment.findOneById<Pick<ILivechatDepartment, '_id' | 'fallbackForwardDepartment'>>(departmentId, { projection: { fallbackForwardDepartment: 1 } })
			: null;
```

File: apps/meteor/ee/server/services/tsconfig.tsbuildinfo
Comment (MarcosSpessatto) [inline apps/meteor/ee/server/services/tsconfig.tsbuildinfo:1]: Committed by mistake maybe?

File: apps/meteor/ee/server/models/raw/LivechatUnit.ts [hunk 8]
```diff
@@ -135,7 +135,7 @@ export class LivechatUnitRaw extends BaseRaw<IOmnichannelBusinessUnit> implement
 		}
 
 		for await (const departmentId of departmentsToSave) {
-			await LivechatDepartment.update(
+			await LivechatDepartment.updateOne(
```
Comment (MarcosSpessatto) [inline apps/meteor/ee/server/models/raw/LivechatUnit.ts:138]: Would not it be possible to use `updateMany` here instead of looping it? 🤔

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

**Review by MarcosSpessatto — APPROVED** [review]

**Review by murtaza98 — APPROVED** [review]

**Pull request merged**